### ISSUE TITLE
node: dist-tag not recorded in package.json

### ISSUE THREAD

**Issue by gradybarrett** [projen/projen#1998]
I've been using prereleases recently and versioning works great, but I noticed that when I set that property the `publishConfig` `tag` property does not get set. I imagine the --tag flag is maybe getting set in a release workflow, but if you're not using workflows, it would be nice for the supplied dist-tag to be tracked in the package.json.

Is there a good reason it's not being tracked in the package.json?

**Comment by mrgrain** [comment]
`publishConfig` is kind of a Quality of Life setting which is less relevant in an automated release workflow like `projen` provides. I guess one way to look at it is that the `tag` option provides is not the same as the one in `publishConfig`. How `projen` ends up publishing the tag is an implementation detail.

Saying that, the expectation management might not be clear here.

**Comment by gradybarrett** [comment]
@mrgrain apologies for the confusion, I had `prerelease` in my head but actually meant `npmDistTag`. Setting the `npmDistTag` property currently does nothing, unless, I assume, you have workflows enabled. That leads to releases not being tagged for folks not using workflows who don't account for this, which is not a great experience.

> ...less relevant in an automated release workflow like projen provides. I guess one way to look at it is that the tag option provides is not the same as the one in publishConfig

Can you help me understand this? Why is it less relevant?

I agree it's an implementation detail. One implementation is `publishConfig.tag` which npm publish respects. Does publib not?

If I set `npmDistTag: 'beta'`, then regardless of the publishing mechanism used, I _expect_ that releases will be tagged with that dist-tag.

**Comment by mrgrain** [comment]
I'm not disagreeing with you at all.

`npmDistTag` is currently tightly coupled to the built-in release mechanism. It would have been more clear to users, if to option would be be `automatedRelease.npmDistTag` or similar. What I meant is that `npmDistTag` currently behaves like that.

Additionally, manual releases aren't well support at the moment. E.g. setting `releaseToNpm: true` and `releaseTrigger: ReleaseTrigger.manual()` doesn't allow to publish to `npm` either.

--- 

It is a very reasonable request to be able to set `publishConfig.tag`. In the meantime you can do this
```
    project.package.addField('publishConfig', {
      tag: 'beta',
    });
```

**Comment by gradybarrett** [comment]
> What I meant is that npmDistTag currently behaves like that.

Gooootcha. 

> I don't know if always setting publishConfig.tag would have any side effects on this.

Release branches came to mind when I was first thinking through this, but in that case each branch has a distinct package.json. dist-tag has a pretty narrow focus, so I can't think of any edge cases. Setting `npmDistTag` is a pretty explicit act, so if I set that, I would expect it to be respected.

Also totally agree with the manual release experience. That's really the backdrop for this issue that I'm trying to figure out on the whole. 😄  Because we don't have workflows, it seems natural that tasks would facilitate that workflow. That means adding tasks for doing a github release if desired, publishing to npm, etc. and then orchestrating all of that as part of the `release` task. https://github.com/projen/projen/pull/1953#issuecomment-1193229032 is highly related.

**Comment by dytyniuk** [comment]
Hello @mrgrain 

I decided to move here since the setting looks more relevant to all those questions I keep asking 😃 

A little context for the rest of the folks here:
* I "jumped on a train" while I was looking for a straightforward way to create a CDK Construct library.
* We use GitLab, not GitHub, hence workflows generated by the projen are no-op for us.
* I thought, that we could set up GitLab CI following the "manual release" feature of the projen.
* While `projen release` does great magic to _cut a release_ and push the chore commit, it does not publish build artifacts to package managers at the end.

For starters, I thought "okay, I could simply use `npm publish` for publishing" 🤷‍♂️ . And then I found out, that the package was published with a version `0.0.0`. Because the `package.json` is always reset before the push and has `version` key `0.0.0`.

I could simply resort to `npm publish dist/js/*.tgz` and reuse a build artifact (and I probably would). The "problem" is this will work if publishing happens in the same run when the release is cut. In case one decides to _manually cut the release_ on their local machine and let CI do the publishing when the git tag is pushed, the `dist` directory simply won't be there.

So, I've spent a couple of hours "disassembling" a set of processes running behind the `projen release` task. And so far I still don't get the reason why `projen` resets the `version` key of `package.json` right before `publish:git`. Since Git is there the source of truth and the latest release Git tag is used to bump the version, I simply cannot foresee any problems.

The `projen release` task flow diagram looks somewhat like the following:

```
+--------+  version      +------------------+
|  bump  +----set-------->   package.json   <---+
+--------+               +--------+---------+   |
                                  |             |
                               version          |              
                                used            |
+---------------------------------+---------+   |
| build                           |         |   |
|                                 |         |   |
| +-----------+                   v         |   |
| |  compile  |                 .jsii       |   |
| +-----------+                   |         |   |
|                                 |         |   |
| +-----------+                   v         |   |
| |  package  |       dist/...<version>.tgz |   |
| +-----------+                             |   |
|                                           |   |
+-------------------------------------------+   |
                                                |
+---------------+       version reset           |
|    unbump     +-------------------------------+
+---------------+

+---------------+
|  publish:git  |
+---------------+
```

I thought I could simply run `projen package` once the tag is pushed. But since the `.jsii` file is ignored, packages are versioned as `0.0.0` if `package` is run separately from the `release`.

Going deeper: `.jsii` file is generated in `projen compile` and ... uses the version from the `package.json` (as you can see on the diagram).

## Local manual releases

So, if one cuts a release manually on their machine, the `dist` directory, `.jsii` file and a version inside `package.json` are lost. In case one (like me) wants to publish packages _after_ the tag is pushed, that can be achieved with the following set of commands:
```shell
# Update the version key inside package.json
$ npm version --no-git-tag-version from-git
# Generate .jsii with the correct version in it
$ projen compile
# Create and publish packages' archives in the dist directory
$ projen package
$ npx publib
```

## Cloud releases
As I have mentioned above, we are using GitLab to host our source code and run CI/CD pipelines. You can find a PoC repository [here](http://gitlab.com/dytyniuk/cdk-construct-lib). I will discuss a desired "cloud release" process today and the repo will get its' CI/CD config soon.

The main question for now is:
> Should CI cut the release after a merge into the `main` branch or a maintainer from his local machine? 🤔

**Comment by mrgrain** [comment]
Okay so now we have a new process in addition to the existing ones:

- 🆕 Start release locally, but finish in the Cloud
- Local manual release
- Cloud release, manually triggered
- Automated Cloud Release

I admit I haven't though much about that new case yet. Can you fill in some gaps for me why someone would want to do this? What needs to happen on a local machine that you can't use `Cloud release, manually triggered`? I can foresee all sorts of problems with this hand-off between local machine and Cloud (with or without version in package.json).

I _really_ like the separation of concerns we have at the moment. Either you let the Cloud deal with it, or not.

--- 

@dytyniuk  Can I ask, for you, is this mostly an issue because GitLab support is bad? And if everything that's available for GitHub were available for GitLab it would solve it for you?

**Comment by dytyniuk** [comment]
@mrgrain I'm not saying, that we _cannot_ start and finalize the release process "in the Cloud". I am going to find this out literally now with my team 😃 

My sincere expectation is that once the Merge Request is merged into the `main` branch, the CI will take care of the release process 🤞 

Even if that configuration is not built by `projen`, having a straightforward action plan would be already a huge gain.

**Comment by mrgrain** [comment]
Understood. I'll try to write something up soon. Thanks for the diagram by the way!

**Comment by dytyniuk** [comment]
Okay, I have got some takeaways from our team talk.

✅ Release process (cutting and publishing) should happen _in the cloud_, i.e. in the GitLab pipeline.

Although the `projen` is a kick-ass tool, TS/JS devs are taken away from their usual package managers (`npm` or `yarn`).
Meaning they have to update `.projenrc.js` manually to have the package in their `package.json`. And even if they add a package via `npm/yarn install`, that will be gone after another `npx projen` run.

That's kinda it 🤷

**Comment by mrgrain** [comment]
> Although the projen is a kick-ass tool, TS/JS devs are taken away from their usual package managers (npm or yarn).
Meaning they have to update .projenrc.js manually to have the package in their package.json. And even if they add a package via npm/yarn install, that will be gone after another npx projen run.

Yes that certainly is a trade-off. I could see a `projen add <package>` happen in future, but it's blocked by #356

**Comment by climbertjh** [comment]
This issue/feature seems to have been dormant for quite awhile now.

Having said that, I've been giving this some thought since I would like to be able to re-create a TypeScript/JavaScript package at any level that has been built using the `projen release` command/operation.

I've only been thinking about this from the point of view of running "local" (i.e. **NOT** trying to make something like this work in the GitHub actions environment).

The high level steps noted above:
1) check-out the level
2) set the package version in `package.json`
3) run the compile
4) run the package

are missing the possibility of pre-compile and post-compile steps.

Also, care would need to be taken by the user not to have un-committed work in flight.

I've been playing around with some additional buildTasks which would run the required steps and restore the working environment.

@mrgrain , @dytyniuk is there any interest in refreshing this discussion?

**Comment by climbertjh** [comment]
I went ahead and created an implementation in my own `projen-utils` package.

See this [(sparse) documentation](https://constructs.dev/packages/@ncfour-us/projen-utils/v/0.1.19/api/TypeScriptESMProjectOptions?lang=typescript#property.buildTagTask) .

If there's some more interest, I will consider submitting a PR.

TL;DR to try it out:
```bash
mkdir tmp-proj; cd tmp-proj
pnpm dlx projen new typescript-esm --from @ncfour-us/projen-utils --buildTagTask
pnpm release
pnpm build:tag v0.0.0
```
( `pnpm release` causes the local git repo to acquire a tag `v0.0.0` which can then be used in the `pnpm build:tag` invocation).
